lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -427,6 +476,126 @@ impl ManagerInner {
+        // - All PFNs must be valid 52-bit page numbers for 4K pages.
+        // - Each overlay kind must appear at most once in each overlay set.
+        self.overlays = payload
+            .overlays
+            .into_iter()
+            .map(|(pfn, set)| -> Result<(Pfn, OverlaySet), OverlayError> {
+                let original_contents = OverlayContents::try_from(
+                    set.original_contents,
+                )
+                .map_err(|len| {
+                    OverlayError::ImportedContentsNotPageSized { pfn, len }
+                })?;
+
+                let pfn = Pfn::new(pfn)
+                    .ok_or(OverlayError::ImportedPfnTooLarge(pfn))?;
+
+                let pending = set
+                    .pending
+                    .into_iter()
+                    .map(|(kind, contents)| -> Result<_, OverlayError> {
+                        if kind == set.active {
+                            return Err(OverlayError::KindInUse(
+                                pfn.into(),
+                                kind.into(),
+                            ));
+                        }
+
+                        let contents = OverlayContents::try_from(contents)
+                            .map_err(|len| {
+                                OverlayError::ImportedContentsNotPageSized {
+                                    pfn: pfn.into(),
+                                    len,
+                                }
+                            })?;
+
+                        Ok((OverlayKind::from(kind), contents))
+                    })
+                    .collect::<Result<BTreeMap<_, _>, OverlayError>>()?;
+
+                Ok((
+                    pfn,
+                    OverlaySet {
+                        original_contents,
+                        active: set.active.into(),
+                        pending,
+                    },
+                ))
+            })
+            .collect::<Result<BTreeMap<Pfn, OverlaySet>, OverlayError>>()?;
+
+        // Now that the tracking tables have been reconstructed, create a set of
+        // overlay page descriptors to return to the caller. The caller will
+        // audit these to make sure they align with other enlightenment state.
+        let pages: Vec<OverlayPage> = self
+            .overlays
+            .iter()
+            .map(|(pfn, set)| {
+                let mut entries = vec![OverlayPage {
+                    manager: Arc::downgrade(manager),
+                    kind: set.active,
+                    pfn: *pfn,
+                }];
+
+                entries.extend(set.pending.keys().map(|kind| OverlayPage {
+                    manager: Arc::downgrade(manager),
+                    kind: *kind,
+                    pfn: *pfn,
+                }));
+
+                entries
+            })
+            .reduce(|mut acc, mut e| {
+                acc.append(&mut e);
+                acc
+            })
```

Comment:
nit, take it or leave it:: I think the `map`/`reduce` could be simplified to `flat_map`?
```suggestion
            .flat_map(|(pfn, set)| {
               std::iter::once(OverlayPage {
                    manager: Arc::downgrade(manager),
                    kind: set.active,
                    pfn: *pfn,
                })
                    .chain(set.pending.keys().map(|kind| OverlayPage {
                        manager: Arc::downgrade(manager),
                        kind: *kind,
                        pfn: *pfn,
                    }))
            })
```

Reply:
Good call--fixed (and added a unit test) in ea80080.